pytorch_lightning/core/optimizer.py

```diff
@@ -0,0 +1,245 @@
+import types
+from abc import ABC
+from argparse import Namespace
+from typing import Any, Callable, Dict, List, Mapping, Optional, Sequence, Tuple, Union
+from weakref import proxy
+
+import torch
+from torch import ScriptModule, Tensor
+from torch.nn import Module
+from torch.optim import SGD
+from torch.optim.optimizer import Optimizer
+
+from pytorch_lightning import _logger as log
+from pytorch_lightning.callbacks import Callback
+from pytorch_lightning.core.grads import GradInformation
+from pytorch_lightning.core.hooks import CheckpointHooks, DataHooks, ModelHooks
+from pytorch_lightning.core.memory import ModelSummary
+from pytorch_lightning.core.saving import ALLOWED_CONFIG_TYPES, PRIMITIVE_TYPES, ModelIO
+from pytorch_lightning.core.step_result import Result
+from pytorch_lightning.utilities import AMPType, rank_zero_warn
+from pytorch_lightning.utilities.device_dtype_mixin import DeviceDtypeModuleMixin
+from pytorch_lightning.utilities.exceptions import MisconfigurationException
+from pytorch_lightning.utilities.parsing import AttributeDict, collect_init_args, get_init_args
+from pytorch_lightning.utilities.xla_device_utils import XLADeviceUtils
+
+TPU_AVAILABLE = XLADeviceUtils.tpu_device_exists()
+
+if TPU_AVAILABLE:
+    import torch_xla.core.xla_model as xm
+
+
+def is_lightning_optimizer(optimizer):
+    return isinstance(optimizer, LightningOptimizer)
+
+
+def do_nothing_closure():
+    return
+
+
+class LightningOptimizer:
+    """
+    This class is used to wrap the user optimizers and handle properly
+    the backward and optimizer_step logic across accelerators, AMP, accumulated_grad_batches
+    """
+    def __init__(self,
+                 optimizer: Optimizer,
+                 accumulate_grad_batches: Optional[int] = None):
+
+        assert accumulate_grad_batches is None or isinstance(accumulate_grad_batches, int)
+        if isinstance(accumulate_grad_batches, int) and accumulate_grad_batches < 1:
+            raise MisconfigurationException(f"accumulate_grad_batches parameters "
+                                            f"{accumulate_grad_batches} should be >= 1")
+
+        self.__dict__ = {k: v for k, v in optimizer.__dict__.items() if k != 'step'}
+
+        # For Horovod
+        if hasattr(optimizer, "skip_synchronize"):
+            self.__class__ = type("Lightning" + optimizer.__class__.__name__, (self.__class__, optimizer.__class__.__bases__[0]), {})
+            self.skip_synchronize = optimizer.skip_synchronize
+            self.synchronize = optimizer.synchronize
+        else:
+            self.__class__ = type("Lightning" + optimizer.__class__.__name__, (self.__class__, optimizer.__class__), {})
+
+        self._trainer = None
+        self._optimizer = optimizer
+        self._accumulate_grad_batches = accumulate_grad_batches
+        self._use_accumulate_grad_batches_from_trainer = accumulate_grad_batches is None
+        # used to find back the step
+        self._lightning_step = self.step
```

Comment:
Is this useful just to locate the step function in the IDE?

Reply:
Which is totally fine, just curious

Reply:
Oh I should remove it. I was using it to overcome Apex overriding step functions.